src/lib/jobrunner.rs

```diff
@@ -0,0 +1,409 @@
+        // Write the JSON to an unnamed temporary file.
+        let json_path = match NamedTempFile::new() {
+            Ok(tfile) => match tfile.into_temp_path().keep() {
+                Ok(p) => {
+                    if fs::write(&p, qj.json_str.as_bytes()).is_err() {
+                        remove_file(p).ok();
+                        return Err(Some(qj));
+                    }
+                    p
+                }
+                Err(_) => return Err(Some(qj)),
+            },
+            Err(_) => return Err(Some(qj)),
+        };
+
+        // We combine the child processe's stderr/stdout and write them to an unnamed temporary
+        // file `stderrout_file`.
+        if let Ok(tempdir) = tempdir() {
+            if let Ok(stderrout_file) = tempfile() {
+                if set_nonblock(stderrout_file.as_raw_fd()).is_ok() {
+                    if let Some(json_path_str) = json_path.to_str() {
+                        let child = match Command::new(qj.path)
+                            .arg(qj.event_type)
+                            .arg(json_path_str)
+                            .current_dir(tempdir.path())
+                            .stderr(process::Stdio::piped())
+                            .stdout(process::Stdio::piped())
+                            .stdin(process::Stdio::null())
+                            .spawn()
+                        {
+                            Ok(c) => c,
+                            Err(e) => {
+                                self.warning(&format!("Can't spawn command: {:?}", e));
+                                return Err(None);
+                            }
+                        };
+
+                        let stderr = child.stderr.as_ref().unwrap();
+                        let stdout = child.stdout.as_ref().unwrap();
+
+                        let stderr_fd = stderr.as_raw_fd();
+                        let stdout_fd = stdout.as_raw_fd();
+                        if let Err(e) =
+                            set_nonblock(stderr_fd).and_then(|_| set_nonblock(stdout_fd))
+                        {
+                            self.warning(&format!(
+                                "Can't set file descriptors to non-blocking: {:?}",
+                                e
+                            ));
+                            return Err(None);
+                        }
+
+                        return Ok(Job {
+                            child,
+                            _tempdir: tempdir,
+                            json_path,
+                            stderrout_file,
+                            stderr_hup: false,
+                            stdout_hup: false,
+                        });
+                    }
+                }
+            }
+        }
+
+        Err(Some(qj))
+    }
+
+    /// After a job has been inserted / removed from `self.running`, this function must be called
+    /// so that `poll()` is called with upto date file descriptors.
```

Comment:
`upto` -> `up to`

Reply:
It turns out it should be "up-to-date". I didn't know that! Fixed in https://github.com/softdevteam/snare/pull/7/commits/5a1af6c45f75b99330e7dd20c9418ca26cadacfc.